docs/dev/virtual_machine.rst

```diff
@@ -0,0 +1,272 @@
+  RegName dst
+  size_t tag
+  size_t num_fields
+  RegName* datatype_fields
+
+Allocate a data type with the tag `tag` using the `num_fields` entries
+from registers datatype_fields. The result is saved to register dst.
+
+### AllocClosure
+**Arguments**:
+  RegName dst
+  size_t clo_index
+  size_t num_freevar
+  RegName* free_vars;
+
+Allocate a closure with the VMFunction at clo_index as
+its code, and the `num_freevar` entries from registers in
+free_vars. The result is saved to register dst.
+
+### GetField
+**Arguments**:
+  RegName dst
+  RegName object
+  size_t field_index
+
+Get the field value with index field_index from object. And saves the result to register dst.
+
+### If
+**Arguments**:
+  RegName if_cond
+  size_t true_offset
+  size_t false_offset
+
+Check if the object at register if_cond is `true` or `false`.
+If true relative jump by `true_offset`, else relative
+jump by `false_offset`.
+
+### Goto
+**Arguments**:
+  size_t pc_offset
+
+Relative unconditional jump by `pc_offset`.
+
+### Invoke
+
+**Arguments**:
+- size_t func_index
+
+Invoke function at `func_index`, consumes the number of arguments contained in the VMFunction's
+arity field, and places return value on stack as top element.
+
+### InvokeClosure
+**Arguments**:
+None
+
+Expects the top value on the stack is a closure. Invokes closure consuming the number of
+arguments declared in the closure_index's VMFunction, and places return value on the stack.
+
+### LoadConst
+**Arguments**:
+  RegName dst
+  size_t const_index
+
+Load the constant at `const_index` from the constant pool. The result is saved to register dst.
+
+# Object Representation
+We use a simple object representation that uses shared pointers and tagging.
+There is a huge space of object representations we can trade off here, but we
+believe micro-optimizing this code has little to no-effect on the end-to-end performance.
+```
+struct ObjectCell {
+  ObjectTag tag;
+  ...
+};
+
+struct Object {
+  std::shared_ptr<ObjectCell> ptr;
+  ...
+}
+```
+See `vm.h` for more details.
+
+Currently we support 3 types of objects: tensors, data types, and closures.
+
+```
+VMObject VMTensor(const tvm::runtime::NDArray& data);
+VMObject VMDatatype(size_t tag, const std::vector<VMObject>& fields);
+VMObject VMClosure(size_t func_index, std::vector<VMObject> free_vars);
+```
+
+# Stack and State
+
+The Relay VM consists of two important stacks, the value stack which acts as the normal call stack
+as used in C/C++/Java/etc, and the frame stack, which contains information about how to resume the
+previous call.
+
+A stack machine is straightforward to implement, register-based virtual machines are more efficient
+in the scalar-world, but we believe in the tensor world that the complexity to performance gain
+tradeoff is not worth it.
+
+We keep track of a set of Relay functions we have called, a pointer into its bytecode, an offset into the byte code (known as the program counter).
+
+```
+struct VirtualMachine {
```

Comment:
Please double check, this is in markdown instead of rst format, you want to turn the doc into rst format

Reply:
@tqchen thanks for the reminder. Fixed in latest commit.